src/backends/interflop-mca/interflop_mca.c

```diff
@@ -122,135 +152,85 @@ static double _mca_rand(void) {
   return tinymt64_generate_doubleOO(&random_state);
 }
 
-static inline double pow2d(int exp) {
-  double res = 0;
-  uint64_t x[1];
-  // specials
-  if (exp == 0)
-    return 1;
-
-  if (exp > 1023) { /*exceed max exponent*/
-    *x = DOUBLE_PLUS_INF;
-    res = *((double *)x);
-    return res;
-  }
-  if (exp < -1022) { /*subnormal*/
-    *x = ((uint64_t)DOUBLE_PMAN_MSB) >> -(exp + DOUBLE_EXP_MAX);
-    res = *((double *)x);
-    return res;
-  }
+static inline double pow2d(int exp) { return ldexp(1.0, exp); }
 
-  // normal case
-  // complement the exponent, shift it at the right place in the MSW
-  *x = (((uint64_t)exp) + DOUBLE_EXP_COMP) << DOUBLE_PMAN_SIZE;
-  res = *((double *)x);
-  return res;
+static inline int32_t rexpq(__float128 q) {
+  binary128 x = {.f128 = q};
+  return x.ieee.exponent - QUAD_EXP_COMP;
 }
 
-static inline uint32_t rexpq(__float128 x) {
-  // no need to check special value in our cases since qnoise will deal with it
-  // do not reuse it outside this code!
-  uint64_t hx, ix;
-  uint32_t exp = 0;
-  GET_FLT128_MSW64(hx, x);
-  // remove sign bit, mantissa will be erased by the next shift
-  ix = hx & QUAD_HX_ERASE_SIGN;
-  // shift exponent to have LSB on position 0 and complement
-  exp = (ix >> QUAD_HX_PMAN_SIZE) - QUAD_EXP_COMP;
-  return exp;
+static inline int32_t rexpd(double d) {
```

Comment:
Could you please briefly comment what each function does, to make code more readable please ?

Reply:
Done.